documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md

```diff
@@ -111,35 +121,170 @@ for fuzzy matches.
 **Finally**, we apply the ranking and sorting parameters, and "highlight" the
 fields that were matched in the results.
 
-The new search controller should allow for using different filters for the first
-step and to not use the full-text search. We should also create a new serializer
-for collection search requests. It should include the common parameters for
-`list` requests, such as `page` and `page_size`, and the parameters for the
-exact matches: `tag`, `creator` and `source`.
+The search controller needs to be updated to extract the first 2 steps into a
+`build_search_query` method. A new `build_collection_query` should be added and
+used when the `collection` parameter is present. This method should create a
+filter query for the relevant field and value.
+
+The pagination and dead link cleanup should be the same for additional search
+queries as for the default search queries.
+
+#### Search request serializer updates
+
+_The initial version of this plan proposed to add new API endpoints and use path
+parameters for the values (e.g., `/image/tag/cat`) to make the URLs more
+readable, easier to share, will be easier to cache or perform cache invalidation
+required by [#1969](https://github.com/WordPress/openverse/issues/1969).
+However, since tags and creator names contain characters that are special for
+path segments (`/`, `?` and `&`) that cannot be properly encoded, this approach
+turned out to be not feasible._
+
+```{note}
+The new query parameters (`collection` and `tag`) will have `unstable__` prefix until this project is
+launched. In the text below, for brevity, the prefix is omitted.
+```
+
+The collections will use the `search` endpoint with `collection` query parameter
+set to `tag`, `creator` or `source`, and the following additional query
+parameters. The `collection` parameter validator will check that the request
+contains the necessary additional parameters: `source` for `collection=source`,
+`creator` and `source` for `collection=creator`, and `tag` for `collection=tag`.
+
+The middleware validates that the values of parameters is not empty, and that
+the `source` parameter is an existing source in the provider store. If the
+values are invalid, the 404 yellow error page is shown.
+
+For the tag route, the new singular `tag` parameter should be used, rather than
+the existing plural `tags`, since we are only presenting a single tag.
+
+The existing `source` and `creator` parameters will be reused, but will be
+parsed differently when `collection` parameter is present: they will only allow
+a single value instead of being split by `,` as it is for the default search. If
+the `source` contains invalid values, such as `source=flickr,europeana`, when
+the `collection` parameter is present, a 400 error with `detail` of "Invalid
+source parameter for source collection: `flickr,europeana`" is returned. If
+possible, a list of valid sources should be returned in the error message.
```

Comment:
Can you clarify the actual implementation details of this? Will it use new `validate_` methods on the serializer, and then for each one check if `collection` is present, and if so, what? Will it throw an error any time a comma is present?

I don't think we can realiably validate this for anything other than source. For all collection parameters, they are passed as-is to the ES query: https://github.com//WordPress/openverse/blob/f56f7310e81ac0266807192803e4a451a8983766/api/api/controllers/search_controller.py#L399-L410

They already do not support multiple values.

Which, again, we need to keep in mind that we cannot reliably split on _any_ character for creator or tag names, neither of which are sanitised to remove commas. What would the validator do if a Flickr creator's name was `Photographer, Esq."? What if the tag is "Taipei, Taiwan"? Will these throw errors? How would a user resolve them? By URL encoding the commas? I'm not sure that will work, they may need to double encode them. Normally, you can pass multiple values for a query parameter by using `param=1,param=2`:

```
> p = new URLSearchParams()
URLSearchParams {}
> p.append("param", "1")
undefined
> p.append("param", "2")
undefined
> p.toString()
'param=1&param=2'
```

Some frameworks require adding `[]` to the front of explicit lists.

We, instead, have chosen to use a comma, which creates significant complication for things like creator. Luckily, we already do not treat creator on the search endpoint as a "listable" input. It is passed as-is in the normal search strategy: 

https://github.com//WordPress/openverse/blob/f56f7310e81ac0266807192803e4a451a8983766/api/api/controllers/search_controller.py#L335-L349

I'm highly sceptical we can reliably validate these inputs, and should instead pass them directly to ES regardless of what is in them. We could validate source to check that it _exists_, so that `flickr,wikimedia_commons` would 404 if passed as `source` with `collection` because there is no `flickr,wikimedia_commons` source. But trying to validate and prevent multiple inputs on tag and creator looks to me like it will create similar issues that we had with the path parameters.

Reply:
I asked a similar question and Olga shared that the intention was to only validate the source, yes. 

https://github.com/WordPress/openverse/pull/3868#discussion_r1513495351

Good call for more specifics here though; I overlooked that this wasn't added in my re-review.

Reply:
I'll add more clarification in the IP itself, but will also add a short explanation here.

For collection view, only two validation steps will be used:

1. if `collection` parameter is present, we check that the relevant parameters are also present:
- if `collection=source`, there is the `source` value in the query
- if `collection=creator`, there are `creator` and `source` values in the query
- if `collection=tag`, there is the `tag` value in the query
2. if collection is `source` or `creator`, we check that the value of the `source` parameter, as is, is present in the list of available creators for the media type. So, `flickr` or `stocksnap` will be valid, but `flick`, `Flickr`, `flickr,stocksnap` will throw an error. We will not split this values by comma or lower-case them. And since we control the creator names, it should be okay. 

As for the `tag` and `creator` parameters, they will be passed to the search controller "as is".

By the way, while trying this out, I realized that our current source validation is not very clear. We take a string of the `source` value, split it by comma, leave all of the values that are names of sources and ignore invalid values. So, `https://api.openverse.engineering/v1/images/?source=flick` ignores the source parameter and returns all of the images. This might be very confusing if the user has made a spelling mistake.

Reply:
Sounds, good, Olga. Thanks for clarifying. It would resolve the issue for me if you update this paragraph to clarify that only the source parameter would ever be validated in any capacity. As it reads now, the creator value would also be validated, but I think that's an artefact of trying to clarify the difference between the normal search route's handling of these two, and the creator parameter getting sidelined a bit.

> By the way, while trying this out, I realized that our current source validation is not very clear. We take a string of the source value, split it by comma, leave all of the values that are names of sources and ignore invalid values. So, https://api.openverse.engineering/v1/images/?source=flick ignores the source parameter and returns all of the images. This might be very confusing if the user has made a spelling mistake.

That sounds like a good new issue, maybe even a "help wanted" one!